### PR TITLE
Preconditions

### LINKED ISSUE #179: Set preconditions based on current account / protocol state
# Precondition RFC

If users want to add preconditions on the account or protocol state, these will most often relate to the _current_ on-chain state of these fields. A very intuitive model is to let users just "use" the current state -- for example, let them access the "current account balance". The mental model is that this balance is a variable (not a constant!) that depends on the current chain. It can be implemented by fetching the current balance from the chain, using it in the prover as the value for the variable, and adding a precondition which fixes the balance to exactly this value. If we wouldn't add the precondition, then the balance would be unconstrained, which would be against the intuition that we prove a computation which uses the current balance as input.

Therefore, I think a good default is to just let users access the balance (and other fields) as the property of some object, and behind the scenes add the necessary precondition. We propose the following API to achieve that:

```ts
// inside method
let balance = this.currentAccount.balance; // automagically adds a precondition
balance.assertLte(100e9); // this method can only be called if current balance <= 100 MINA
```

Note that to use this API intuitively, one doesn't need to understand the concept of a precondition at all. It's enough to be aware of the `this.currentAccount` property. The returned `balance` is a plain `UInt64` variable; the added precondition stays completely hidden. Similarly, we propose to have a `this.currentProtocolState` property which exposes protocol state preconditions. On other parties, we would have corresponding `party.currentX` properties.

However, advanced users may want to declare arbitrary preconditions. For example, in the above example, it makes sense to actually _not_ use the current balance, but just to add a precondition directly. We propose the following API:

```ts
Party.assertBetween(this.accountPrecondition.balance, 0, 100e9);
```

This works for any party, not just the `this` party. It mutates `this.accountPrecondition.balance`, which _is_ the precondition (represented as an object `{lower: UInt64, upper: UInt64}`.

The last example achieves the same as the first one -- it constrains the balance to be less than 100 MINA -- but without the fragility that the current balance needs to stay unchanged until the transaction is accepted. However, to use an API like this, you do have to understand the concept of a precondition. That's why it's OK to use more jargon ("precondition"), and make this a bit less discoverable (have to use `Party`), and let it not do any magic. Similar to `this.accountPrecondition`, we propose to expose `this.protocolStatePrecondition`, which would just be shortcuts to access `party.body.accountPrecondition` and `party.body.protocolStatePrecondition`.

In addition to `Party.assertBetween`, there is also `Party.assertEquals` (already implemented), and we could likewise add `Party.assertBelow` and `Party.assertAbove`.

Finally, you may want to _reference the current state_ when declaring an arbitrary precondition. For example, possibly you want to restrict the nonce to an interval of the form `[currentNonce, currentNonce + tolerance]`. Or you may want to restrict it to be exactly `currentNonce + 2`, because you already know that there'll be 2 earlier transactions which increment the same nonce. For this final use case, we propose the following API:

```ts
Party.assertEquals(this.accountPrecondition.nonce, this.currentAccount.nonce.add(2));
```

This is the precisely same API as before, just making use of `this.currentAccount`. The logic would be that any explicitly set precondition would override any magically inserted precondition. So, in this case, nonce == currentNonce would be overriden to nonce == currentNonce + 2.

## TODO list
* [ ] Implement the `this.current*` API and, for other parties, `party.current*`.
* [ ] Implement `this.*Precondition` and `party.*Precondition`.
* [x] Implement `Party.assertBetween`, `Party.assertEquals`
* [ ] Fetch all account preconditions from chain
* [ ] Fetch all protocol state preconditions from chain
* [ ] Create unit-tests which confirm that 
  - setting/fetching all preconditions is possible
  - set preconditions show up in the JSON transaction

### PR DESCRIPTION
**Pull request by mitschabaude**
- Closes #179
- Adds example of using the new API: https://github.com/o1-labs/snarkyjs/blob/dab7ed9aa93fe95fb88f1b184195de252046704c/src/examples/simple_zkapp.ts#L39
- Adds a method for checking deep equality of two party subtypes (where "party subtype" is defined very broadly to be future-proof): https://github.com/o1-labs/snarkyjs/pull/207/commits/5bf05cdc3341b5159498daabb11b1613f313d1a7. This is used to check if the user manually modified a precondition, which counts as explicitly constraining the value, to avoid the error that we throw when it's not constrained.

TODOs:
* [ ] change integration test so that it tests fetching from the graphql URI, and uses preconditions, and ideally is broken up into zkapps that test different features
* [x] investigate broken prover and decide if it should be fixed as part of this issue
* [ ] some precondition fields aren't implemented yet because the logic to parse them from an encoded string is missing. this could be done here or in a follow-up PR

### CODE REVIEW

File: src/lib/fetch.ts [hunk 1]
```diff
@@ -218,34 +245,6 @@ function stringifyAccount(account: FlexibleAccount): FetchedAccount {
   };
 }
 
-async function checkResponseStatus(
```
Comment (mitschabaude) [inline src/lib/fetch.ts:221]: this code block was just moved to the bottom of the file. I often move unimportant / helper-type code to the bottom, and  the exported API / most important parts to the top, because I think this makes code more readable

File: src/lib/precondition.ts [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+  return { account: Account(party), network: Network(party) };
+}
+
+// note: please keep the two precondition implementations separate
+// so we can add customized fields easily
+
+function Network(party: Party): Network {
+  // TODO there should be a less error-prone way of typing js layout
+  // e.g. separate keys list and value object, so that we can access by key
+  let layout = (jsLayout as any).Party.layout[0].value.layout[9].value.layout[0]
```
Comment (mitschabaude) [inline src/lib/precondition.ts:36]: the `jsLayout` provides the necessary info to produce `precondition` interfaces of the right structure at runtime

File: src/lib/precondition.ts [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+function Account(party: Party): Account {
+  // TODO there should be a less error-prone way of typing js layout
+  // e.g. separate keys list and value object, so that we can access by key
+  let layout = (jsLayout as any).Party.layout[0].value.layout[9].value.layout[1]
+    .value as Layout;
+  let context = getPreconditionContextExn(party);
+  return preconditionClass(layout, 'account', party, context);
+}
+
+function preconditionClass(
```
Comment (mitschabaude) [inline src/lib/precondition.ts:51]: this function is recursive! it produces all the `preconditions.account` and `preconditions.network` fields

File: src/lib/precondition.ts [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+  preconditionContexts.set(party, {
+    read: new Set(),
+    constrained: new Set(),
+    vars: {},
+    isSelf,
+  });
+}
+
+function assertPreconditionInvariants(party: Party) {
+  let context = getPreconditionContextExn(party);
```
Comment (mitschabaude) [inline src/lib/precondition.ts:191]: this is the function that throws an error if the user used `.get()` on a account / network field, without explicitly adding a precondition. calling `get` would have added an entry to `context.read`. adding a precondition would have added an entry to `context.constrained`.

File: src/snarky.d.ts [hunk 5]
```diff
@@ -417,6 +428,15 @@ declare class Bool {
   toBoolean(): boolean;
 
   /* static members */
+  /**
+   * The constant [[`Bool`]] that is `true`.
+   */
+  static true: Bool;
```
Comment (mitschabaude) [inline src/snarky.d.ts:434]: I added `Bool.true` and `Bool.false`, because in the end it won the community vote :D

File: src/lib/int.ts [hunk 6]
```diff
@@ -1,15 +1,18 @@
...
+import { Circuit, Field, Types } from '../snarky';
 import { CircuitValue, prop } from './circuit_value';
 
 export { UInt32, UInt64, Int64 };
 
-class UInt64 extends CircuitValue {
+class UInt64 extends CircuitValue implements Types.UInt64 {
   @prop value: Field;
   static NUM_BITS = 64;
+  type = 'UInt64' as const;
```
Comment (mitschabaude) [inline src/lib/int.ts:9]: I added this field to `UInt64`, to make the type conform to the interface `{ value: Field; type: "UInt64" }`, which is used to signify `UInt64`s in the autogenerated party types.
the intention is to make the `UInt64` interface distinct from the `UInt32` one, to get advanced types in `precondition.ts`to work. this also fixes a couple of bug sources, where `UInt32` could be passed into functions that expected `UIn64` and vice versa

File: src/lib/int.ts [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
 
   constructor(value: Field) {
     super();
     this.value = value;
+    // TODO type breaks Circuit.if for some reason
+    delete (this as any).type;
```
Comment (mitschabaude) [inline src/lib/int.ts:15]: we actually can't keep the `.type` property on the class instance.. it's just for typescript

File: src/lib/party.ts [hunk 8]
```diff
@@ -482,10 +483,26 @@ type LazyControl = Control | LazySignature | LazyProof;
...
+  constructor(body: Body, authorization = {} as LazyControl, isSelf = false) {
     this.body = body;
+    this.authorization = authorization;
+    let { account, network } = Precondition.preconditions(this, isSelf);
+    this.account = account;
+    this.network = network;
+    this.isSelf = isSelf;
+  }
+
+  static clone(party: Party) {
```
Comment (mitschabaude) [inline src/lib/party.ts:502]: Before adding this, `Party`s were just cloned with `cloneCircuitValue` directly. I changed it because the new `account` and `network` fields aren't safely clone-able, due to how they use closures / are attached to a party instance. This new cloning will lose precondition-related metadata attached to a party (i.e. which account / network fields were accessed / constrained). this seems appropriate, because cloning a party also doesn't magically add it to the current transaction

**Review by mitschabaude — COMMENTED** [review]
Some comments to help review:

File: src/lib/int.ts (hunk 6 above)
Comment (bkase) [reply src/lib/int.ts:9]: Can we use a phantom type here instead of actually polluting the object at runtime? https://dev.to/busypeoples/notes-on-typescript-phantom-types-kg9 You can put the type parameter on the interface to avoid people having to see it when they use Uint64 .

File: src/lib/party.ts (hunk 8 above)
Comment (bkase) [reply src/lib/party.ts:502]: 👍

File: src/snarky.d.ts (hunk 5 above)
Comment (bkase) [reply src/snarky.d.ts:434]: Nice

File: src/lib/int.ts (hunk 6 above)
Comment (bkase) [reply src/lib/int.ts:9]: See below, I think I solved it nicely

**Review by bkase — CHANGES_REQUESTED** [review]
I don’t know if I believe in the WeakMap over storing secret things on the object.
The benefit is you can easily see everything associated with that entity by looking at the class. When there are WeakMap’s floating around, you have to know where to look to find the extra associated data. 

Also, popular and well-thought-out libraries do this (like React), so I don’t think it’s controversial.

What is your reason for reaching for weakmaps here?

If we do end up sticking with WeakMaps, having just one for all the secret things at least mitigates the “hard to find” associated data problem a bit. Rather than one for each kind of associated datum.

**Comment by mitschabaude** [comment]
> I don’t know if I believe in the WeakMap over storing secret things on the object. (...) Also, popular and well-thought-out libraries do this (like React), so I don’t think it’s controversial.

Ok, this is good feedback. I might have overreached with the intention to move all secret props away. Specifically, I'm not sure if I can defend the move of `SmartContract._executionState` to a weakmap, because that prop mainly gets used *internally* by the `SmartContract` class. In that case, I guess it's more natural to make the prop `private` so it's removed from the TS type / doesn't pollute intellisense, but still present at runtime.

> What is your reason for reaching for weakmaps here?

> If we do end up sticking with WeakMaps, having just one for all the secret things at least mitigates the “hard to find” associated data problem a bit. Rather than one for each kind of associated datum.

I do think there are cases where weakmaps are better than extra class props, and it's also better to not have them in one place. For example, the `precondition.ts` module stores metadata associated with each `Party` instance, in a weakmap. I like the weakmap approach in this case because it makes the metadata book-keeping be localized to the module that is concered with it - `precondition.ts`. I think keeping modules encapsulated helps with understanding and maintaining the code. In this case, seeing a map of precondition-related metadata being initialized on the `Party` class in `party.ts` would be more confusing, because there's no other code in `party.ts` concerned with that metadata. It would mix concerns of the `Party` class with concerns of the `precondition` module.

An alternative would be to secretly set a property `party._preconditionData = ...` inside `precondition.ts`, but I think this is just not as clean, not as type-able etc as maintaining a weakmap for that data.

The same reasoning applies for the `@state` logic that's now encapsulated in `state.ts`. We used to store `_states` and `_layout()` fields on `SmartContract` subclasses from inside the `@state` logic. I think it's strictly less confusing and more maintainable now that we have a state-related weakmap inside `state.ts`, and not mess with the `SmartContract`s from that module.

Since only `precondition.ts` / `state.ts` ever need to be concerned with their respective metadata, I don't see the value of having a central weakmap which combines their stuff. Also, the keys and values of that weakmap would be a mix of completely different things if we do that, making accurate typing harder.

TL;DR, I think both of thos weakmaps help with encapsulation, code clarity and strict typing!

@bkase, what do you think of a middle-ground approach where we don't try to avoid private props in principle, but we use weakmaps if a module wants to associate data with objects that originate in a different module? I would revert the `_executionState` weakmap in that case.

File: src/snarky/parties-leaves.ts [hunk 9]
```diff
@@ -12,8 +12,8 @@ export {
 
 export { toJson, toJsonLeafTypes, toFields, toFieldsLeafTypes };
 
-type UInt64 = { value: Field };
-type UInt32 = { value: Field };
+type UInt64 = { value: Field; type: 'UInt64' };
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/snarky/parties-leaves.ts:15]: Unfortunately doesn't seem to solve my use case.. I need `UInt64` to not extend `UInt32` and vice versa.

I tried the following snippet:
```ts
type FieldLike<N> = { value: Field };
type UInt64 = FieldLike<{ type: 'UInt64' }>;
type UInt32 = FieldLike<{ type: 'UInt32' }>;
type X = UInt32 extends UInt64 ? 'not working' : 'it works!';
```
For me, `X` is of type `"not working"`..

[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/snarky/parties-leaves.ts:15]: I think my original solution can be made slightly better by making the type prop optional:

```ts
type UInt64 = { value: Field; type?: 'UInt64' };
type UInt32 = { value: Field; type?: 'UInt32' };
```
that way, they don't extend each other and we can just instantiate them without the `type` field at runtime.

it's still not perfect because the `type` field shows up in intellisense

File: src/lib/fetch.ts [hunk 10]
```diff
@@ -317,6 +347,163 @@ function addCachedAccountInternal(
   };
 }
 
+async function fetchLastBlock(graphqlEndpoint = defaultGraphqlEndpoint) {
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/lib/fetch.ts:350]: very nice, I did that and it became much faster!

File: src/lib/fetch.ts [hunk 11]
```diff
@@ -317,6 +347,163 @@ function addCachedAccountInternal(
...
+      totalCurrency,
+      slot,
+      slotSinceGenesis,
+      nextEpochData,
+      stakingEpochData,
+    },
+  },
+}: FetchedBlock): NetworkValue {
+  return {
+    snarkedLedgerHash: Field.zero, // TODO
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/lib/fetch.ts:460]: done 👍🏻

File: src/lib/precondition.ts [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+  for (let preconditionPath of context.read) {
+    // check if every precondition that was read was also contrained
+    if (context.constrained.has(preconditionPath)) continue;
+
+    // check if the precondition was modified manually, which is also a valid way of avoiding an error
+    let precondition = getPath(party.body.preconditions, preconditionPath);
+    let dummy = getPath(dummyPreconditions, preconditionPath);
+    if (!circuitValueEquals(precondition, dummy)) continue;
+
+    // we accessed a precondition field but not constrained it explicitly - throw an error
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/lib/precondition.ts:234]: I added a file with unit tests. Please check whether you think it covers everything: https://github.com/o1-labs/snarkyjs/pull/207/commits/899d14ce8d349c7d5d7876104f28e26f6c57cb10

File: src/snarky/parties-leaves.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bkase) [reply src/snarky/parties-leaves.ts]: That's weird that they extend each other, shows that I do not understand typescript's type system enough.

I think this is good with them optional -- since it appears in intellisense can you use some kind of prefix so people know to avoid it: like `_type` or `$type` (I'm not sure if we have a standard for this yet)

**Comment by bkase** [comment]
Renaming the `type?` to `_type?` (or something) + refactoring that WeakMap for executionState and then I'll approve 💯

**Review by bkase — APPROVED** [review]
